controllers/subscriptions.go

```diff
@@ -259,9 +261,48 @@ func GetVendorCsvNames(cli client.Client, kind odfv1alpha1.StorageKind) ([]strin
 		csvNames = []string{OcsSubscriptionStartingCSV, RookSubscriptionStartingCSV, NoobaaSubscriptionStartingCSV,
 			PrometheusSubscriptionStartingCSV, RecipeSubscriptionStartingCSV}
 
-		if isProvider, err = isProviderMode(cli); !isProvider {
+		isProvider, err = isProviderMode(cli)
+		if err != nil {
+			return csvNames, err
+		}
+
+		if !isProvider {
 			csvNames = append(csvNames, OcsClientSubscriptionStartingCSV, CSIAddonsSubscriptionStartingCSV)
 		}
+
+		// In provider mode, upgrades of the ocs-client-operator and csiaddons are managed by the provider, not the odf-operator.
+		// This can result in these operators lagging behind the odf-operator, with different CSV versions.
+		// Therefore, we need to fetch the CSV name from the operator deployment.
+		// We are only fetching the ocs-client-operator CSV name and ignoring the csiaddons
+		// Because the ocs-client-operator is essential when enabling provider mode as we need to bring it up.
+
+		// Fetch the CSV name from the client operator deployment and append it to the csv list
+		if isProvider {
+			// get ocs-client-operator deployment with label
+			deployments := &appsv1.DeploymentList{}
+			err = cli.List(context.TODO(), deployments, &client.ListOptions{
```

Comment:
Preferably let the caller pass a context and use it.
It does effect, cancelability, logging and stack traces

Reply:
Actually we need a lot of changes for that in multiple places, Lets leave that for later.